text/0111-auto-resource-detection.md

```diff
@@ -0,0 +1,178 @@
+detects a resource, the resources will be merged according to the Merge
+interface already defined in the specification, i.e. the earliest matched
+resource's attributes will take precedence. Each detector may be run in
+parallel, but to ensure deterministic results, the resources must be merged in
+the order the detectors were added.
+
+A default implementation of a detector that reads resource data from the
+`OTEL_RESOURCE` environment variable will be included in the SDK. The
+environment variable will contain of a list of key value pairs, and these are
+expected to be represented in a format similar to the [W3C
+Correlation-Context](https://github.com/w3c/correlation-context/blob/master/correlation_context/HTTP_HEADER_FORMAT.md#header-value),
+except that additional semi-colon delimited metadata is not supported, i.e.:
+`key1=value1,key2=value2`. If the user does not specify any resource, this
+detector will be run by default.
+
+Custom resource detectors related to specific environments (e.g. specific cloud
+vendors) must be implemented as packages separate to the core SDK, and users
+will need to import these separately.
+
+## Internal details
+
+As described above, the following will be added to the Resource SDK
+specification:
+
+- An interface for "detectors", to retrieve resource information
+- Specification for a global function to merge resources returned by a set of
+  detectors
+- Details of the "from environment variable" detector implementation as
+  described above
+- Specification that default detection (from environment variable) runs once on
+  startup, and is used by all tracer & meter providers by default if no custom
+  resource is supplied
+
+### Usage
+
+The following example in Go creates a tracer and meter provider that uses
+resource information automatically detected from AWS or GCP:
+
+Assumes a dependency has been added on the `otel/api`, `otel/sdk`,
+`otel/awsdetector`, and `otel/gcpdetector` packages.
+
+```go
+resource, _ := sdkresource.Detect(ctx, 5 * time.Second, awsdetector.ec2, gcpdetector.gce)
+tp := sdktrace.NewProvider(sdktrace.WithResource(resource))
+mp := push.New(..., push.WithResource(resource))
+```
+
+### Components
+
+#### Detector
+
+The `Detector` interface will simply contain a `Detect` function that returns a
+Resource.
+
+The `Detect` function should contain a mechanism to timeout and cancel the
+request. If a detector is not able to detect a resource, it must return an
+uninitialized resource such that the result of each call to `Detect` can be
+merged.
+
+#### Global Function
+
+The SDK will also provide a global `Detect` function. This will take a timeout
+duration and a set of detectors that should be run and merged in order as
+described in the intro, and return a resource.
+
+### Error Handling
+
+In the case of one or more detectors raising an error, there are two reasonable
+options:
+
+1. Ignore that detector, and continue with a warning (likely meaning we will
```

Comment:
Please compare https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/error-handling.md

Reply:
I don't think there's anything here that contradicts those guidelines given that detection code runs on initialization.

Usually we would "assume that users would prefer to lose telemetry data rather than have the library significantly change the behavior of the instrumented application", but given that this code runs on initialization, this seems to come under pont no. 2: "The API or SDK may _fail fast_ and cause the application to fail on initialization".

Having the detection code fail fast, and leaving it up to the user to handle this if desired seems reasonable?

Reply:
Sorry, that was a rather lazy and vague comment from me 😄 I guess what struck me as odd here is in what detail you describe error handling here. I think most of the specification does at most say "must not throw exceptions" and otherwise this document is the only thing that has more details.